DOC: Update arraycreation

cooperrc opened the pull request:
This PR is part of the Google Season of Docs with @melissawm and @mattip 

This PR updates the basics.creation.rst to reflect some ongoing discussions with `dtype` assignments and it builds the document as an [explanation-doc](https://documentation.divio.com/explanation/). 

The highlighted changes:

* 5 creation methods -> 6 creation methods (I included replication, joining, and mutating as array creation methods)
* I numbered the different methods
* I divided NumPy array creation routines into:
  1. 1D arrays (e.g. linspace/arange)
  2. 2D arrays (e.g. eye, diag)
  3. ND arrays (e.g. zeros, ones)
* I put in sphinx links for as many functions mentioned as I could

I could use some help with intersphinx to reference SciPy IO and Pandas. 

Closes #17710

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

cooperrc committed:
rewriting basics.creation

cooperrc committed:
halfway through first draft basics.creation

cooperrc committed:
first draft for array creation rewrite

cooperrc committed:
correct some broken links

cooperrc committed:
correct typo in code format

melissawm commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,138 +9,353 @@ Array creation
 
-Note that there are some subtleties regarding the last usage that the user
-should be aware of that are described in the arange docstring.
+Note: there are some subtleties regarding ``dtype``. In the second
```
Maybe here we could use a `.. note::` directive? Otherwise, the text could flow better without the colon, I think.

melissawm commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,138 +9,353 @@ Array creation
-should be aware of that are described in the arange docstring.
+Note: there are some subtleties regarding ``dtype``. In the second
+example, the ``dtype`` is defined. In the third example, the array is
+``dtype = float`` to accomodate the step size of ``0.1``. 
```
PEP8

```suggestion
``dtype=float`` to accomodate the step size of ``0.1``. 
```

melissawm commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,138 +9,353 @@ Array creation
-The advantage of this creation function is that one can guarantee the
-number of elements and the starting and end point, which arange()
-generally will not do for arbitrary start, stop, and step values.
+The advantage of this creation object is that you guarantee the
+number of elements and the starting and end point. The previous
+``arange(start, stop, step)`` will not include the value ``stop``.
```
Any reason to change from function to object in the text above?

I think it might be helpful to include a note here about [expected `arange` behaviour](https://github.com/numpy/numpy/issues/17342) with non-integer steps.

cooperrc replied:
> Any reason to change from function to object in the text above?

I missed that mention of "function". I originally wrote all my notes as referring to "functions" instead of the preexisting vocabulary "objects".

I'll change ~~function~~ -> "object"

> I think it might be helpful to include a note here about [expected `arange` behaviour](https://github.com/numpy/numpy/issues/17342) with non-integer steps.

I hadn't run into this problem before, thanks for suggesting the note! This is a great place to discuss this behavior. 

Why does `arange` sometimes include the `stop` value? Is it a roundoff issue?

mattip replied:
> Any reason to change from function to object in the text above?

I think the question was the opposite: use `function` not `object`. Although everything in python is an object, it is more precise to refer to the routines as `functions`.

> Why does arange sometimes include the stop value? Is it a roundoff issue?

Yes, and it is a common source of new user confusion.

melissawm commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,138 +9,353 @@ Array creation
+
+:func:`numpy.diag` can define either a square 2D array with given values along
+the diagonal _or_ if given a 2D array returns a 1D array that is
+only the diagonal elements. The two array creations can be helpful while
+doing linear algebra as such::
```
```suggestion
only the diagonal elements. The two array creation functions can be helpful while
doing linear algebra, as such::
```

melissawm commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,138 +9,353 @@ Array creation
+The 2D array creation objects e.g. :func:`numpy.eye`, :func:`numpy.diag`, and :func:`numpy.vander`
+define properties of special matrices represented as 2D arrays. 
+
+``eye(n,m)`` defines a 2D identity matrix. The elements where i=j are 1
```
```suggestion
``np.eye(n, m)`` defines a 2D identity matrix. The elements where i=j are 1
```

melissawm commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,138 +9,353 @@ Array creation
+ array([[1., 0., 0.],
+        [0., 1., 0.],
+        [0., 0., 1.]])
+ >>> eye(3,5)
```
PEP8

```suggestion
 >>> eye(3, 5)
```

melissawm commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,138 +9,353 @@ Array creation
+only the diagonal elements. The two array creations can be helpful while
+doing linear algebra as such::
+ 
+ >>> diag([1,2,3])
```
PEP8

```suggestion
 >>> diag([1, 2, 3])
```

melissawm commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,138 +9,353 @@ Array creation
+ array([[1, 0, 0],
+       [0, 2, 0],
+       [0, 0, 3]])
+ >>> diag([1,2,3],1)
```
PEP8

```suggestion
 >>> diag([1, 2, 3], 1)
```

melissawm commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,138 +9,353 @@ Array creation
+ >>> diag(a)
+ array([1, 4])
+
+``vander(x,n)`` defines a Vandermonde matrix as a 2D NumPy array. Each column
```
PEP8

```suggestion
``vander(x, n)`` defines a Vandermonde matrix as a 2D NumPy array. Each column
```

melissawm commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,138 +9,353 @@ Array creation
+Once you have created arrays, you can replicate, join, or mutate those
+existing arrays to create new arrays. When you assign an array or its
+elements to a new variable, you have to explicitly :func:`numpy.copy` the array,
+otherwise the variable is a view into the original array. Consider the
```
Might be worth linking to [copies and views](https://numpy.org/devdocs/user/quickstart.html#copies-and-views)

cooperrc replied:
Good idea, thanks

melissawm commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,138 +9,353 @@ Array creation
+        [ 0.,  0., -3.,  0. ],
+        [ 0.,  0.,  0., -4. ]])
+
+Other routines use similar syntax to join ND arrays, check the
```
```suggestion
Other routines use similar syntax to join ND arrays. Check the
```

melissawm commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,138 +9,353 @@ Array creation
 
  HDF5: h5py
  FITS: Astropy
+ # ?? anything else??
```
There could be many including Zarr, netCDF and others, but I think this can be done in a follow-up.

cooperrc replied:
removed the `# anything else` and I'll leave for another time

melissawm commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,138 +9,353 @@ Array creation
-simple format then one can write a simple I/O library and use the numpy
-fromfile() function and .tofile() method to read and write numpy arrays
+simple format then one can write a simple I/O library and use the NumPy
+fromfile() function and .tofile() method to read and write NumPy arrays
```
```suggestion
`fromfile()` function and `.tofile()` method to read and write NumPy arrays
```

melissawm commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,138 +9,353 @@ Array creation
+Many Python libraries, including SciPy, Pandas, and OpenCV, can create,
+operate on, and work with NumPy arrays. 
+..
+  These libraries could benefit from a few examples, but it may be
```
Agreed - this can be done later.

melissawm commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,138 +9,353 @@ Array creation
-Custom Binary Formats
----------------------
+
+More generic ascii files can be read using `SciPy
+io<https://docs.scipy.org/doc/scipy/reference/io.html>`_ and `Pandas
+<https://pandas.pydata.org/>`_.
+..
+  intersphinx links for io and Pandas?
```
I think Pandas can be as it is. This should work:
```suggestion
More generic ASCII files can be read using `scipy.io` and `Pandas
<https://pandas.pydata.org/>`_.
```

melissawm reviewed (changes requested):
Hi @cooperrc ! Thanks for this.

I only have a few comments (typos, PEP8 fixes and smaller considerations). I don't have any issues with the content, seems ok to me.

The largest comment is noted below: should we go with "ndarray" to be consistent with the rest of the docs? I think so.

cooperrc commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,138 +9,353 @@ Array creation
-This section will not cover means of replicating, joining, or otherwise
-expanding or mutating existing arrays. Nor will it cover creating object
-arrays or structured arrays. Both of those are covered in their own sections.
+You can use these methods to create ND arrays or :doc:`structured_arrays`.
+This document will cover general methods for ND array creation. 
```
Yes, I should change ND array to ndarray for docs consistency. Thanks for pointing that out. I like ND array, but it can't be the only occurrence in the docs.

cooperrc commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,138 +9,353 @@ Array creation
+
+:func:`numpy.arange` creates arrays with regularly incrementing values.
+Check the docstring for complete information and examples. A few
+examples are shown: ::
```
I'm fine dropping the colon in these "for example:" endings. 

I had left a ":" to end the explanation and move to code, but it's probably not necessary.

eric-wieser replied:
`foo::` both starts a code block _and_ outputs a single colon

cooperrc committed:
Apply suggestions from code review - @melissawm

PEP8 corrections and some other mistakes. thanks @melissawm!

Co-authored-by: Melissa Weber Mendonça <melissawm@gmail.com>

charris commented:
CircleCI failure looks legitimate.

cooperrc commented:
> CircleCI failure looks legitimate.

Should I update to current NumPy master. You think that would fix it? I'd be surprised if these edits to the basics.creation.rst would affect CircleCI.

cooperrc committed:
remove comments on L347 and 356, fix link on L22

cooperrc committed:
remove comments on L347 and 356, fix link on L22

cooperrc commented:
> The failures are in rendering the docs:
> 
> ```
> /home/circleci/repo/doc/source/user/basics.creation.rst:347: WARNING: Unexpected indentation.
> /home/circleci/repo/doc/source/user/basics.creation.rst:356: WARNING: Unexpected indentation.
> /home/circleci/repo/doc/source/user/basics.creation.rst:22: WARNING: unknown document: structured_arrays
> ```
Thanks
OK, L347 and L356 were comments. I'll delete those. 

L22 is a reference that didn't work. Should work now.

cooperrc committed:
remove typo on L355 section title error

cooperrc committed:
change ND array -> ndarray docs consistent vocab

cooperrc committed:
added np.* to numpy functions and dtypes

cooperrc commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,138 +9,353 @@ Array creation
+    unsigned c: [4294967293 4294967293 4294967293] uint32
+    >>> c_signed32 = a - b.astype(int32)
+    >>> print('signed c:', c_signed32, c_signed32.dtype)
+    signed c: [-3 -3 -3] int64
```
updated the functions to use `np.<...>` in each NumPy-specific function

mattip commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,55 +9,101 @@ Array creation
-see if it works! (The Python Way).
-
-Examples: ::
+1) Converting Python structures to NumPy Arrays
```
```suggestion
1) Converting Python sequences to NumPy Arrays
```

mattip commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,55 +9,101 @@ Array creation
- >>> x = np.array([[1,2.0],[0,0],(1+1j,3.)]) # note mix of tuple and lists,
-     and types
- >>> x = np.array([[ 1.+0.j, 2.+0.j], [ 0.+0.j, 0.+0.j], [ 1.+1.j, 3.+0.j]])
+NumPy arrays can be defined using Python structures such as lists and
```
```suggestion
NumPy arrays can be defined using Python sequences such as lists and
```

mattip commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,55 +9,101 @@ Array creation
- >>> x = np.array([[ 1.+0.j, 2.+0.j], [ 0.+0.j, 0.+0.j], [ 1.+1.j, 3.+0.j]])
+NumPy arrays can be defined using Python structures such as lists and
+tuples. Lists and tuples are defined using ``[...]`` and ``(...)``,
+respectively. Lists and tuples can define ndarray creation:
```
This is not creating a unorderd list. Is another `:` needed?
```suggestion
respectively. Lists and tuples can define ndarray creation::
```

mattip commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,55 +9,101 @@ Array creation
+respectively. Lists and tuples can define ndarray creation:
+* a list of numbers will create a 1D array, 
+* a list of lists will create a 2D array, 
+* further nested lists will create ndarrays
```
```suggestion
* further nested lists will create higher-dimensional arrays. In general,
  any array object is called an **ndarray** in NumPy.
```

mattip commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,55 +9,101 @@ Array creation
+
+NumPy has over 40 built-in functions for creating arrays as laid
+out in the :ref:`Array creation routines <routines.array-creation>`.
+There are three categories of NumPy array creation objects:
+1) 1D arrays
+2) 2D arrays
+3) ndarrays
```
Not rendered as a list. Also, technically they all are ndarrays...
```suggestion
There are three categories of NumPy array creation objects::
1) 1D arrays (or 1D ndarrays)
2) 2D arrays (or 2D ndarrays)
3) general ndarrays (just called ndarray)
```

mattip commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,55 +9,101 @@ Array creation
+1 - 1D array creation objects
+-----------------------------
+
+The 1D array creation objects e.g. :func:`numpy.linspace` and
```
```suggestion
The 1D array creation functions e.g. :func:`numpy.linspace` and
```

mattip commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,55 +9,101 @@ Array creation
+
+The 1D array creation objects e.g. :func:`numpy.linspace` and
+:func:`numpy.arange` generally need at least two inputs, ``start`` and
+``stop``, where ``start`` is the first value in the array and ``stop``
+specifies the value of the last element in the array. 
```
The description is misleading: stop is not the value of the last element in the array.
```suggestion
``stop``. 
```

mattip commented on doc/source/user/basics.creation.rst:
```diff
@@ -66,43 +112,188 @@ examples will be given here: ::
+        [27,  9,  3,  1],
+        [64, 16,  4,  1]])
+ 
+3 - ndarray creation objects
```
```suggestion
3 - general ndarray creation functions
```

mattip commented on doc/source/user/basics.creation.rst:
```diff
@@ -66,43 +112,188 @@ examples will be given here: ::
+number of elements and the starting and end point. The previous
+``arange(start, stop, step)`` will not include the value ``stop``.
+
+2 - 2D array creation objects
```
```suggestion
2 - 2D array creation functions
```

mattip commented on doc/source/user/basics.creation.rst:
```diff
@@ -66,43 +112,188 @@ examples will be given here: ::
+2 - 2D array creation objects
+-----------------------------
+
+The 2D array creation objects e.g. :func:`numpy.eye`, :func:`numpy.diag`, and :func:`numpy.vander`
```
```suggestion
The 2D array creation functions e.g. :func:`numpy.eye`, :func:`numpy.diag`, and :func:`numpy.vander`
```

mattip commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,55 +9,101 @@ Array creation
+2) 2D arrays
+3) ndarrays
+
+1 - 1D array creation objects
```
I would drop the use of numbered lists. You are already inside point (2) Intrinsic NumPy array creation objects, so the added numbers are confusing, and the right navbar clearly specifies the layered hierarchy..
```suggestion
1 - 1D array creation functions
```

mattip commented on doc/source/user/basics.creation.rst:
```diff
@@ -66,43 +112,188 @@ examples will be given here: ::
+3 - ndarray creation objects
+-----------------------------
+
+The ndarray creation objects e.g. :func:`numpy.ones`, :func:`numpy.zeros`, and
```
```suggestion
The ndarray creation functions e.g. :func:`numpy.ones`, :func:`numpy.zeros`, and
```

mattip commented on doc/source/user/basics.creation.rst:
```diff
@@ -66,43 +112,188 @@ examples will be given here: ::
+-----------------------------
+
+The ndarray creation objects e.g. :func:`numpy.ones`, :func:`numpy.zeros`, and
+:func:`default_rng.random` define arrays based upon the desired
```
Link does not work as written. Maybe this?
```suggestion
`default_rng().random <random.Generator.random>` define arrays based upon the desired
```

mattip commented on doc/source/user/basics.creation.rst:
```diff
@@ -66,43 +112,188 @@ examples will be given here: ::
+
+:func:`numpy.indices` will create a set of arrays (stacked as a one-higher
+dimensioned array), one per dimension with each representing variation in that
+dimension.  An example illustrates much better than a verbal description: ::
```
```suggestion
dimension: ::
```

mattip commented on doc/source/user/basics.creation.rst:
```diff
@@ -66,43 +112,188 @@ examples will be given here: ::
  array([[[0, 0, 0], [1, 1, 1], [2, 2, 2]], [[0, 1, 2], [0, 1, 2], [0, 1, 2]]])
 
 This is particularly useful for evaluating functions of multiple dimensions on
 a regular grid.
```
What about an example of `default_rng().random` ?

cooperrc replied:
adding one now. Thanks for catching :+1:

mattip commented on doc/source/user/basics.creation.rst:
```diff
@@ -66,43 +112,188 @@ examples will be given here: ::
+existing arrays to create new arrays. When you assign an array or its
+elements to a new variable, you have to explicitly :func:`numpy.copy` the array,
+otherwise the variable is a view into the original array. Consider the
+following example, ::
```
```suggestion
following example::
```

mattip commented on doc/source/user/basics.creation.rst:
```diff
@@ -66,43 +112,188 @@ examples will be given here: ::
+
+This is the most common case of large array creation. The details depend
+greatly on the format of data on disk. This section gives
+general pointers on how to handle various formats.
```
Maybe a links to the IO howto somewhere in here?

cooperrc replied:
got it, thanks!

mattip commented on doc/source/user/basics.creation.rst:
```diff
@@ -114,33 +305,48 @@ convert are those formats supported by libraries like PIL (able to read and
+6) Use of special library functions (e.g., SciPy, Pandas, and OpenCV)
+=====================================================================
 
+NumPy is a fundamental library in the Python Scientific Computing stack. 
+Many Python libraries, including SciPy, Pandas, and OpenCV, can create,
+operate on, and work with NumPy arrays. 
```
```suggestion
NumPy is the fundamental library for array containers in the Python Scientific Computing
stack. Many Python libraries, including SciPy, Pandas, and OpenCV, use NumPy ndarrays
as the common format for data exchange, These libraries can create,
operate on, and work with NumPy arrays. 
```

mattip reviewed (commented):
A "few" comments :). This is a great improvement, thanks.

cooperrc committed:
@mattip review changes

Great changes by @mattip. I like the change from "object" -> "function"

Co-authored-by: Matti Picus <matti.picus@gmail.com>

eric-wieser commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,55 +9,101 @@ Array creation
-shape. The default dtype is float64. ::
+ >>> a1D = np.array([1, 2, 3, 4])
+ >>> a2D = np.array([[1, 2], [3, 4]])
+ >>> a3D = np.array([[[1, 2], [3, 4]],
+                    [[5, 6], [7, 8]]
+                    ])
```
```suggestion
 >>> a3D = np.array([[[1, 2], [3, 4]],
                     [[5, 6], [7, 8]]])
```

eric-wieser commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,55 +9,101 @@ Array creation
 
- >>> np.zeros((2, 3))
- array([[ 0., 0., 0.], [ 0., 0., 0.]])
+When you use :func:`numpy.array` to define a new array, consider the
+:doc:`dtype <basics.types>` of the elements in the array. You should
+specify the ``dtype`` for elements in the array. This feature gives you
```
These two sentences are rather repetitive.
```suggestion
When you use :func:`numpy.array` to define a new array, you should
consider the :doc:`dtype <basics.types>` of the elements in the array,
which can be specified explicitly. This feature gives you
```

eric-wieser commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,55 +9,101 @@ Array creation
-docstring for complete information on the various ways it can be used. A few
-examples will be given here: ::
+ >>> a = np.array([127, 128, 129], dtype=np.int8)
+ >>> print(a)
+ [ 127 -128 -127 ]
```
It might be better to not encourage `print` here:
```suggestion
 >>> a
 array([ 127, -128, -127], dtype=int8)
```

eric-wieser commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,55 +9,101 @@ Array creation
+
+NumPy has over 40 built-in functions for creating arrays as laid
+out in the :ref:`Array creation routines <routines.array-creation>`.
+There are three categories of NumPy array creation objects::
```
```suggestion
These functions can be split into roughly three categories, based on the
dimension of the array they create::
```

cooperrc replied:
adding to next commit

eric-wieser commented on doc/source/user/basics.creation.rst:
```diff
@@ -66,43 +112,188 @@ examples will be given here: ::
+shape. The default dtype is ``float64``. ::
+
+ >>> np.zeros((2, 3))
+ array([[ 0., 0., 0.], 
+        [ 0., 0., 0.]])
```
```suggestion
 array([[0., 0., 0.], 
        [0., 0., 0.]])
```

eric-wieser commented on doc/source/user/basics.creation.rst:
```diff
@@ -66,43 +112,188 @@ examples will be given here: ::
+tuple or list. 
+
+:func:`numpy.zeros` will create an array filled with 0 values with the specified
+shape. The default dtype is ``float64``. ::
```
```suggestion
shape. The default dtype is ``float64``::
```

Else this displays as "`float64`. :"

eric-wieser commented on doc/source/user/basics.creation.rst:
```diff
@@ -66,43 +112,188 @@ examples will be given here: ::
+         [0., 0.]]])
+
+:func:`numpy.ones` will create an array filled with 1 values. It is identical to
+``zeros`` in all other respects as such, ::
```
```suggestion
``zeros`` in all other respects as such::
```

cooperrc committed:
added default_rng() example and some links

cooperrc committed:
fixing merge 20200126

cooperrc committed:
Update doc/source/user/basics.creation.rst

reformat 3D array L40

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

cooperrc committed:
Apply suggestions from code review

thanks @eric-wieser for catching some formatting typos.

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

cooperrc committed:
Merge remote-tracking branch 'upstream/master' into update_arraycreation

cooperrc committed:
working on updates and trying to build

cooperrc commented:
@melissawm and @mattip thanks for the comments/suggestions. I apologize for letting this sit, the Holidays were much more hectic than I imagined. 

I made most of the suggested changes, but then hit `checks were not successful` again. I'm having some trouble building NumPy with `pip install .`. I hit the import error:
```
We have compiled some common reasons and troubleshooting tips at:

    https://numpy.org/devdocs/user/troubleshooting-importerror.html
```

Am I missing a step in the build process? or doing it wrong?

mattip commented:
More information needed. What error exactly are you seeing?

mattip commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,55 +9,100 @@ Array creation
+ >>> a1D = np.array([1, 2, 3, 4])
+ >>> a2D = np.array([[1, 2], [3, 4]])
+ >>> a3D = np.array([[[1, 2], [3, 4]],
+                     [[5, 6], [7, 8]]])
```
Did you mean to indent this block?

cooperrc replied:
> Did you mean to indent this block?

Yes, I was following the convention I found in the `basics.creation.rst` where the code was indented with a space in the examples:

```
::

\s>>> example = 1
```

Should they have a 4-space indent for PEP-8?

mattip replied:
That would be nicer, I thought the failure to build the docs that pointed to line 34 was because there is no indentation. I missed the single space :(

cooperrc replied:
> That would be nicer, I thought the failure to build the docs that pointed to line 34 was because there is no indentation. I missed the single space :(

I missed it at first too, looking at `absolute_beginners.rst` the code is indented with two space-indents `\s\s`. 

Q: Should there be a Sphinx indentation standard in the docs?

mattip commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,55 +9,100 @@ Array creation
+then you need to specify the ``dtype`` while you create the array. 
+
+2) Intrinsic NumPy array creation functions
+=========================================
```
underline too short

melissawm replied:
There are other places where this is failing, you can see them [in the CircleCI build output](https://app.circleci.com/pipelines/github/numpy/numpy/5898/workflows/55401f06-bc08-4f31-9b71-dc527e2508ec/jobs/17985)

cooperrc replied:
thanks, updating now

cooperrc commented:
> More information needed. What error exactly are you seeing?

My process: 
1. create fresh Python 3.8 conda environment
2. install cython
3. `pip install .` in NumPy source folder

Here is the output from `pip install .`
```
  Created wheel for numpy: filename=numpy-1.21.0.dev0+566.g49cbaf96f-cp38-cp38-linux_x86_64.whl size=17034457 sha256
=11a020022e59157a7507556add9a3f1857508b6290038ac24b86052463664c09                                                   
  Stored in directory: /tmp/pip-ephem-wheel-cache-f2hpe7hs/wheels/2f/4d/d7/a2e907151ccb709c255da24ca6e22caa300bdb551
9815b1ac2                                                                                                           
Successfully built numpy                                                                                            
Installing collected packages: numpy                                                                                
Successfully installed numpy-1.21.0.dev0+566.g49cbaf96f
```

Then, I open a Python interpreter and get the following error with `import numpy`
```
Traceback (most recent call last):                                                                                  
  File "/home/ryan/.conda/envs/doc-fresh-38/lib/python3.8/site-packages/numpy/core/__init__.py", line 22, in <module
>                                                                                                                   
    from . import multiarray                                                                                        
  File "/home/ryan/.conda/envs/doc-fresh-38/lib/python3.8/site-packages/numpy/core/multiarray.py", line 12, in <modu
le>                                                                                                                 
    from . import overrides                                                                                         
  File "/home/ryan/.conda/envs/doc-fresh-38/lib/python3.8/site-packages/numpy/core/overrides.py", line 7, in <module
>                                                                                                                   
    from numpy.core._multiarray_umath import (                                                                      
ImportError: /home/ryan/.conda/envs/doc-fresh-38/lib/python3.8/site-packages/numpy/core/_multiarray_umath.cpython-38
-x86_64-linux-gnu.so: undefined symbol: cblas_sgemm                                                                 
                                                                                                                    
During handling of the above exception, another exception occurred:                                                 
                                                                                                                    
Traceback (most recent call last):                                                                                  
  File "<stdin>", line 1, in <module>                                                                               
  File "/home/ryan/.conda/envs/doc-fresh-38/lib/python3.8/site-packages/numpy/__init__.py", line 150, in <module>   
    from . import core                                                                                              
File "/home/ryan/.conda/envs/doc-fresh-38/lib/python3.8/site-packages/numpy/core/__init__.py", line 48, in <module
>
    raise ImportError(msg)
ImportError: 

IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the numpy C-extensions failed. This error can happen for
many reasons, often due to issues with your setup or how NumPy was
installed.

We have compiled some common reasons and troubleshooting tips at:

    https://numpy.org/devdocs/user/troubleshooting-importerror.html

Please note and check the following:

  * The Python version is: Python3.8 from "/home/ryan/.conda/envs/doc-fresh-38/bin/python"
  * The NumPy version is: "1.21.0.dev0+566.g49cbaf96f"

and make sure that they are the versions you expect.
Please carefully study the documentation linked above for further help.

Original error was: /home/ryan/.conda/envs/doc-fresh-38/lib/python3.8/site-packages/numpy/core/_multiarray_umath.cpy
thon-38-x86_64-linux-gnu.so: undefined symbol: cblas_sgemm 
```

mattip commented:
> `_multiarray_umath.cpython-38
-x86_64-linux-gnu.so: undefined symbol: cblas_sgemm`

Something is off with the BLAS implementation found when building. I assume you are on Linux? If `-ldd multiarray_umath.cpython-38-x86_64-linux-gnu.so` is showing a system library, can you update? You could do something like `python tools/openblas_support.py` (which will download and unpack the OpenBLAS used in CI), it should tell you where it put the files, let's say that is `<path-reported>`. Then you can create a `site.cfg` with the content
```
[openblas]
libraries = openblas
library_dirs = <path-reported>/lib
include_dirs = <path-reported>/include
runtime_library_dirs = <path-reported>/lib
```

If I didn't mislead you, now when you build numpy and `ldd` the c-extension module, it should use that openblas.

eric-wieser commented on doc/source/user/basics.creation.rst:
```diff
@@ -66,43 +111,212 @@ examples will be given here: ::
+         [1., 1.],
+         [1., 1.]]])
+
+:func:`default_rng.random` will create an array filled with random
```
I think this will need to be something like:
```suggestion
The :meth:`~numpy.random.Generator.random` method of the result of :func:`default_rng` will create an array filled with random
```
for the link to work

eric-wieser commented on doc/source/user/basics.creation.rst:
```diff
@@ -66,43 +111,212 @@ examples will be given here: ::
+ array([[0.77395605, 0.43887844, 0.85859792],
+        [0.69736803, 0.09417735, 0.97562235]])
+ >>> default_rng(42).random((2,3,2))
+ array([[[0.77395605, 0.43887844],
+		 [0.85859792, 0.69736803],
+		 [0.09417735, 0.97562235]],
+
+		 [[0.7611397 , 0.78606431],
+		 [0.12811363, 0.45038594],
+		 [0.37079802, 0.92676499]]])
```
Indents are screwy here

cooperrc replied:
in vim, they looked fine. I updated to get them aligned again.

thanks!

eric-wieser replied:
You've mixed tabs and spaces - there should be no tabs in this file.

eric-wieser commented on doc/source/user/basics.creation.rst:
```diff
@@ -114,33 +328,49 @@ convert are those formats supported by libraries like PIL (able to read and
-simple format then one can write a simple I/O library and use the numpy
-fromfile() function and .tofile() method to read and write numpy arrays
+simple format then one can write a simple I/O library and use the NumPy
+`fromfile()` function and `.tofile()` method to read and write NumPy arrays
```
I don't think these will link properly as is. Either try to fix the links, or change to double backticks.

cooperrc replied:
changed to ``

thanks

cooperrc commented:
> > `_multiarray_umath.cpython-38 -x86_64-linux-gnu.so: undefined symbol: cblas_sgemm`
> 
> Something is off with the BLAS implementation found when building. I assume you are on Linux? If `ldd multiarray_umath.cpython-38-x86_64-linux-gnu.so` is showing a system library, can you update? You could do something like `python tools/openblas_support.py` (which will download and unpack the OpenBLAS used in CI), it should tell you where it put the files, let's say that is `<path-reported>`. Then you can create a `site.cfg` with the content
> 
> ```
> [openblas]
> libraries = openblas
> library_dirs = <path-reported>/lib
> include_dirs = <path-reported>/include
> runtime_library_dirs = <path-reported>/lib
> ```
> 
> If I didn't mislead you, now when you build numpy and `ldd` the c-extension module, it should use that openblas.

The `site.cfg` + `openblas_support.py` installation worked, thanks! no errors with `import numpy` now.

cooperrc committed:
fixed heading underlines and some links

cooperrc committed:
Apply suggestions from code review

fixing link to default_rng

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

melissawm commented on doc/source/user/basics.creation.rst:
```diff
@@ -66,43 +111,212 @@ examples will be given here: ::
+The 2D array creation functions e.g. :func:`numpy.eye`, :func:`numpy.diag`, and :func:`numpy.vander`
+define properties of special matrices represented as 2D arrays. 
+
+``np.eye(n, m)`` defines a 2D identity matrix. The elements where i=j are 1
```
I don't know if this is necessary, feel free to ignore.

```suggestion
``np.eye(n, m)`` defines a 2D identity matrix. The elements where i=j (row index and column index are equal) are 1
```

melissawm commented on doc/source/user/basics.creation.rst:
```diff
@@ -66,43 +111,212 @@ examples will be given here: ::
+The ndarray creation functions e.g. :func:`numpy.ones`,
+:func:`numpy.zeros`, and `default_rng().random
+<random.Generator.random>` define arrays based upon the desired shape.
+The  ndarray creation functions can create arrays with any dimension by
```
```suggestion
The ndarray creation functions can create arrays with any dimension by
```

melissawm commented on doc/source/user/basics.creation.rst:
```diff
@@ -66,43 +112,211 @@ examples will be given here: ::
+ >>> print('a = ', a, 'b = ', b)
+ a =  [1 2 3 4 5 6] b =  [2 3]
+
+For more information and examples ook at :ref:`Copies and Views
+<quickstart.copies-and-views>.
```
```suggestion
For more information and examples look at :ref:`Copies and Views
<quickstart.copies-and-views>`.
```

cooperrc committed:
Apply suggestions from code review - @melissawm

adding @melissawm comments to changes

Co-authored-by: Melissa Weber Mendonça <melissawm@gmail.com>

melissawm commented:
On my side things are ok, just pending the indentation issues. Thanks, @cooperrc !

cooperrc committed:
reword ndarray creation distinctions

cooperrc committed:
fixed a merge with github

cooperrc commented:
> On my side things are ok, just pending the indentation issues. Thanks, @cooperrc !

Thanks @melissawm!

regarding indentation, do you mean the outputs or the code blocks? I think I caught the indentation issues we discussed above.

mattip commented:
Checkout the CI failure on cirecleCI build -> details. It is complaining about indentation and line 34.

cooperrc committed:
L34 indented and L94 literal block removed

cooperrc committed:
L34 remove newline to try to fix indent make error

cooperrc commented:
> Checkout the CI failure on cirecleCI build -> details. It is complaining about indentation and line 34.

Thanks, I see that. I don't know why its complaining about the indent for the list. I tried making it one line in the last commit since it doesn't like my indents (third time's a charm).

cooperrc commented:
> > Checkout the CI failure on cirecleCI build -> details. It is complaining about indentation and line 34.
> 
> Thanks, I see that. I don't know why its complaining about the indent for the list. I tried making it one line in the last commit since it doesn't like my indents (third time's a charm).

L34 indentation was changed to a single line. It fixed the circleci build test. Looks like we are ready to merge.

melissawm commented on doc/source/user/basics.creation.rst:
```diff
@@ -9,55 +9,101 @@ Array creation
-==============================
+NumPy arrays can be defined using Python sequences such as lists and
+tuples. Lists and tuples are defined using ``[...]`` and ``(...)``,
+respectively. Lists and tuples can define ndarray creation::
+* a list of numbers will create a 1D array, 
+* a list of lists will create a 2D array, 
+* further nested lists will create higher-dimensional arrays. In general, any array object is called an **ndarray** in NumPy.
```
There's a tiny formatting issue here:
```suggestion
respectively. Lists and tuples can define ndarray creation:

* a list of numbers will create a 1D array, 
* a list of lists will create a 2D array, 
* further nested lists will create higher-dimensional arrays. In general, any array object is called an **ndarray** in NumPy.
```

cooperrc committed:
Update doc/source/user/basics.creation.rst

@melissawm caught a formatting issue for a list, `::` -> `:` and extra line added

Co-authored-by: Melissa Weber Mendonça <melissawm@gmail.com>

eric-wieser commented on doc/source/user/basics.creation.rst:
```diff
@@ -114,33 +334,49 @@ convert are those formats supported by libraries like PIL (able to read and
+functions can read and parse these files line-by-line. NumPy has two
+standard routines for importing a file with delimited data :func:`numpy.loadtxt`
+and :func:`numpy.genfromtxt`. These functions have more involved use cases in
+:doc:`how-to-io`. A simple example given a ``simple.csv``::
+
+ $ cat simple.csv
```
```suggestion
:doc:`how-to-io`. A simple example given a ``simple.csv``:

.. code-block:: bash

 $ cat simple.csv
```
else this will be treated as python

eric-wieser commented on doc/source/user/basics.creation.rst:
```diff
@@ -66,43 +113,216 @@ examples will be given here: ::
+
+There are a number of routines to join existing arrays e.g. :func:`numpy.vstack`,
+:func:`numpy.hstack`, and :func:`numpy.block`. Here is an example of joining four 2-by-2
+arrays into a 4-by-4 array using ``block`` ::
```
```suggestion
arrays into a 4-by-4 array using ``block``::
```

eric-wieser commented on doc/source/user/basics.creation.rst:
```diff
@@ -66,43 +113,216 @@ examples will be given here: ::
+ >>> a = np.array([1, 2, 3, 4, 5, 6])
+ >>> b = a[:2]
+ >>> b += 1
+ >>> print('a = ', a, 'b = ', b)
+ a =  [2 3 3 4 5 6] b =  [2 3]
```
May as well make this look tidy:
```suggestion
 >>> print('a =', a, '; b =', b)
 a = [2 3 3 4 5 6]; b = [2 3]
```
although perhaps easiest to read as
```suggestion
 >>> a
 array([2, 3, 3, 4, 5, 6])
 >>> b
 array([2, 3])
```

charris committed:
MAINT: Update doc/source/user/basics.creation.rst

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

charris committed:
MAINT: Update doc/source/user/basics.creation.rst

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

charris committed:
MAINT: Update doc/source/user/basics.creation.rst

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

charris commented:
Thanks @cooperrc .

cooperrc commented:
> Thanks @cooperrc .

Thanks to the careful reviewers @melissawm, @mattip, and @eric-wieser!! I'm very proud of our work.
